libqpdf/QPDF.cc

```diff
@@ -1942,18 +1942,14 @@ QPDF::readObjectAtOffset(
 void
 QPDF::resolve(QPDFObjGen const& og)
 {
-    if (isCached(og) && !isUnresolved(og)) {
-        // We only need to resolve unresolved objects
-        return;
-    }
+    // resolve is (and must only be) called with unresolved objects.
 
-    // Check object cache before checking xref table.  This allows us
-    // to insert things into the object cache that don't actually
-    // exist in the file.
     if (this->m->resolving.count(og)) {
```

Comment:
I'd get rid of `m->resolving` entirely and just have throw a logic error if resolve is called on something that's resolved.